corehq/messaging/smsbackends/infobip/models.py

```diff
@@ -47,25 +53,80 @@ def get_opt_out_keywords(cls):
             return False
-        # TODO: Add other exceptions here
 
-    def _send_text_message(self, config, to, msg):
+    def _send_omni_failover_message(self, config, to, msg, headers):
         payload = {
             'destinations': [{'to': {'phoneNumber': to}}],
             'scenarioKey': config.scenario_key,
-            'whatsApp': {'text': msg.text}
+            'viber': {'text': msg.text},
+            'line': {'text': msg.text},
+            'voice': {'text': msg.text},
+            'sms': {'text': msg.text}
+        }
+        if is_whatsapp_template_message(msg.text):
+            try:
+                parts = get_template_hsm_parts(msg.text)
+            except WhatsAppTemplateStringException:
+                msg.set_system_error(SMS.ERROR_MESSAGE_FORMAT_INVALID)
+            payload['whatsApp'] = {
+                'templateName': parts.template_name,
+                'language': parts.lang_code,
+                'templateData': parts.params
+            }
+        else:
+            payload['whatsApp'] = {'text': msg.text}
+
+        url = f'https://{config.personalized_subdomain}.{INFOBIP_DOMAIN}/omni/1/advanced'
+        response = requests.post(url, json=payload, headers=headers)
+        return response.content
+
+    def _send_sms(self, config, to, msg, headers):
+        payload = {
+            'messages': [{
+                'from': config.reply_to_phone_number,
+                'destinations': [{'to': to}],
+                'text': msg.text
+            }]
         }
+        url = f'https://{config.personalized_subdomain}.{INFOBIP_DOMAIN}/sms/2/text/advanced'
+        response = requests.post(url, json=payload, headers=headers)
+        return response.content
+
+    def get_all_templates(self):
         headers = {
-            'Authorization': f'App {config.auth_token}',
+            'Authorization': f'App {self.config.auth_token}',
             'Content-Type': 'application/json',
             'Accept': 'application/json'
         }
-        url = f'https://{INFOBIP_HOST}/omni/1/advanced'
-        response = requests.post(url, json=payload, headers=headers)
-        return response.content
+        url = f'https://{self.config.personalized_subdomain}.{INFOBIP_DOMAIN}' \
+            f'/whatsapp/1/senders/{self.config.reply_to_phone_number}/templates'
+        response = requests.get(url, headers=headers)
+        templates = []
+        try:
+            templates = json.loads(response.content).get('templates')
+        except Exception:
+            raise
+        return templates
+
+    @classmethod
+    def generate_template_string(cls, template):
+        """From the template JSON returned by Infobip, create the magic string for people to copy / paste
+        """
+
+        template_text = template.get("body", "")
+        num_params = template_text.count("{") // 2  # each parameter is bracketed by {{}}
+        parameters = ",".join([f"{{var{i}}}" for i in range(1, num_params + 1)])
```

Comment:
Two less brackets can't hurt
```suggestion
        parameters = ",".join(f"{{var{i}}}" for i in range(1, num_params + 1))
```

Reply:
Thanks. I didn't forget about this suggestion and all other for these branch. I just couldn't commit them since this branch is closed. I'll include them in my next branch.